lib/src/facts/linux/processor_resolver.cc

```diff
@@ -13,65 +14,155 @@ using namespace std;
                 return true;
             }
-            string key = line.substr(0, pos);
-            boost::trim(key);
-            string value = line.substr(pos + 1);
-            boost::trim(value);
 
             if (key == "processor") {
                 // Start of a logical processor
                 id = move(value);
                 if (!have_counts) {
-                    ++result.logical_count;
+                    ++data.logical_count;
                 }
             } else if (!id.empty() && key == "model name") {
                 // Add the model for this logical processor
-                result.models.emplace_back(move(value));
+                data.models.emplace_back(move(value));
             } else if (!have_counts && key == "physical id" && cpus.emplace(move(value)).second) {
                 // Couldn't determine physical count from sysfs, but CPU topology is present, so use it
-                ++result.physical_count;
+                ++data.physical_count;
+            }
+            return true;
+        });
+    }
+
+    void processor_resolver::add_power_cpu_data(data& data, bool have_counts, std::string const& root)
+    {
+        unordered_set<string> cpus;
+        string id;
+        lth_file::each_line(root + "/proc/cpuinfo", [&](string& line) {
+            string key, value;
+            if (!split_line(line, key, value)) {
+                return true;
+            }
+
+            if (key == "processor") {
+                // Start of a logical processor
+                id = move(value);
+                if (!have_counts) {
+                    ++data.logical_count;
+                }
+            } else if (!id.empty() && key == "cpu") {
+                // Add the model for this logical processor
+                data.models.emplace_back(move(value));
+            } else if (key == "clock" && data.speed == 0) {
+                // Parse out the processor speed (in MHz)
+                string speed;
+                if (lth_util::re_search(value, boost::regex("(\\d+).*MHz"), &speed)) {
+                    data.speed = stoi(speed) * static_cast<int64_t>(1000);
+                }
             }
             return true;
         });
 
+    }
+
+    void processor_resolver::add_cpu_data(data& data, std::string const& root)
+    {
+        unordered_set<string> cpus;
+        lth_file::each_subdirectory(root + "/sys/devices/system/cpu", [&](string const& cpu_directory) {
+            ++data.logical_count;
+            string id = lth_file::read((path(cpu_directory) / "/topology/physical_package_id").string());
+            boost::trim(id);
+            if (id.empty() || cpus.emplace(move(id)).second) {
+                // Haven't seen this processor before
+                ++data.physical_count;
+            }
+            return true;
+        }, "^cpu\\d+$");
+
+        bool have_counts = data.logical_count > 0;
+        
+        if (architecture_type(data, root) == GENERIC) {
```

Comment:
It looks like both `architecture_type` and `add_*_cpu_data` read the contents of `/proc/cpuinfo` - might be nice to read it only once and pass in the output string.

Reply:
Yes, I'm going to use that instead of root at the end since that's the whole reason the root is passed in there. Good idea!